cache-explainer.md

```diff
@@ -0,0 +1,133 @@
+
+## Proposed API
+
+```webidl
+partial interface MLContext {
+  Promise<sequence<DOMString>> listGraphs();
+  Promise<MLGraph> loadGraph(DOMString key);
+  Promise<undefined> saveGraph(DOMString key, MLGraph graph);
+  undefined deleteGraph(DOMString key);
+};
+```
+
+1. **`listGraphs()`**: Returns a list of keys for all cached graphs.
+2. **`loadGraph(key)`**: Loads a cached graph associated with the given key.
+3. **`saveGraph(key, graph)`**: Saves the provided graph under the specified key.
+4. **`deleteGraph(key)`**: Deletes the cached graph associated with the given key.
+
+### A note on persistence
+
+A graph may be evicted from the cache due to storage pressure or browser/platform updates which render previously compiled graphs invalid. Developers should consider the level of durability to be somewhere between IndexedDB and the HTTP cache. [For specification purposes, reuse the [Storage standard concepts](https://storage.spec.whatwg.org/#model) as applicable.]
+
+### Input and output descriptors
+
+A JS ML framework, such as ONNX Runtime Web, may need to know the input and output operands info (name, shape and data type) to construct input and output tensors for an inference session. The input and output operands info is known if users pass the source model, e.g. ONNX model. With model cache, user may only pass the model key, the framework needs to fetch the input and output operands info from an `MLGraph`. It would be necessary to expose the `inputDescriptors` and `outputDescriptors` internal slots of `MLGraph` interface.
+
+```webidl
+partial interface MLGraph {
+  record<USVString, MLOperandDescriptor> inputDescriptors;
+  record<USVString, MLOperandDescriptor> outputDescriptors;
+};
+```
+
+## Considered alternatives
+
+### Combined build and save
+
+A separate `saveGraph` API might introduce overhead on some native ML frameworks, such as ONNX Runtime, because its implementation may need to hold the source model in the memory and recompile the source model when user code calls `saveGraph`.
+
+An alternative consideration is to have a `buildAndSave` method. The implementation can just compile the graph once and drop the source model after the compilation.
+
+```webidl
+partial interface MLGraphBuilder {
+  Promise<MLGraph> buildAndSave(MLNamedOperands outputs, DOMString key);
+};
+```
+
+### Explicit vs implicit API
+
+>GPU shader caching is implicit, however the difference is that a shader program is a small input and so it's easy for the site to regenerate the shader so the browser can hash it to compare with the cache. ML models on the other hand are large because of the weights. Loading all the weights just to discover that a cached version of the model is available would be a waste of time and resources. (via [comment](https://github.com/webmachinelearning/webnn/issues/807#issuecomment-2608135598))
```

Comment:
Remove the ">" here.

Reply:
Fixed in https://github.com/webmachinelearning/webnn/pull/862/commits/1e1bc1c23991d2169f8c59d7c9de20dd6454b63e